File: artifacts/discord-bot/modules/hosting.py
"""
HOSTING MODULE — KaluxHost Themed
Commands: plans, uptime, status, support, ticket, node
Both prefix and slash variants.
"""
import datetime
import discord
from discord import app_commands
from discord.ext import commands

from main.config import COLOR_BRAND, BOT_NAME
from main.utils.embeds import brand, success, warn


PLANS = [
    {"name": "⚡ Starter",   "ram": "2 GB",  "cpu": "1 vCore",   "storage": "20 GB SSD",  "price": "$2.99/mo", "color": 0x57F287},
    {"name": "🚀 Pro",       "ram": "4 GB",  "cpu": "2 vCores",  "storage": "50 GB SSD",  "price": "$5.99/mo", "color": COLOR_BRAND},
    {"name": "💎 Elite",     "ram": "8 GB",  "cpu": "4 vCores",  "storage": "100 GB SSD", "price": "$9.99/mo", "color": 0xFEE75C},
    {"name": "👑 Dedicated", "ram": "16 GB+","cpu": "8+ vCores", "storage": "250 GB SSD", "price": "Custom",   "color": 0xED4245},
]

SERVICES = [
    ("🟢", "Web Panel"),
    ("🟢", "Game Servers"),
    ("🟢", "VPS / VDS Nodes"),
    ("🟢", "Discord Bot"),
    ("🟢", "Billing System"),
    ("🟢", "DDoS Protection"),
]


def _plans_embed() -> discord.Embed:
    embed = discord.Embed(
        title="🖥️ KaluxHost — Hosting Plans",
        description="Premium hosting with **99.9% uptime guarantee**.\nAll plans include DDoS protection & 24/7 support.",
        color=COLOR_BRAND,
        timestamp=datetime.datetime.now(datetime.timezone.utc),
    )
    for p in PLANS:
        embed.add_field(
            name=p["name"],
            value=f"RAM: `{p['ram']}`\nCPU: `{p['cpu']}`\nDisk: `{p['storage']}`\nPrice: **{p['price']}**",
            inline=True,
        )
    embed.set_footer(text=f"{BOT_NAME} | Visit our website to order!")
    return embed


def _status_embed() -> discord.Embed:
    embed = brand("📊 KaluxHost — Service Status")
    embed.description = "\n".join(f"{icon} **{name}** — Operational" for icon, name in SERVICES)
    return embed


class Hosting(commands.Cog, name="Hosting"):
    """🖥️ KaluxHost hosting commands."""

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    # ── !plans ────────────────────────────────────────────────────────────────

    @commands.command(name="plans", aliases=["pricing", "packages"])
    async def plans_cmd(self, ctx: commands.Context):
        """View KaluxHost hosting plans."""
        await ctx.reply(embed=_plans_embed(), mention_author=False)

    # ── !status ───────────────────────────────────────────────────────────────

    @commands.command(name="status", aliases=["services"])
    async def status_cmd(self, ctx: commands.Context):
        """Check KaluxHost service status."""
        await ctx.reply(embed=_status_embed(), mention_author=False)

    # ── !uptime ───────────────────────────────────────────────────────────────

    @commands.command(name="uptime")
    async def uptime_cmd(self, ctx: commands.Context):
        """Check bot and service uptime."""
        delta   = datetime.timedelta(milliseconds=round(self.bot.latency * 1000))
        uptime  = datetime.datetime.now(datetime.timezone.utc) - self.bot.start_time
        embed   = brand("📡 Uptime Status")
        embed.add_field(name="🤖 Bot Online Since", value=f"<t:{int(self.bot.start_time.timestamp())}:R>", inline=True)
        embed.add_field(name="📶 API Latency",       value=f"`{round(self.bot.latency*1000)}ms`",           inline=True)
        embed.add_field(name="🟢 Services",           value="All systems operational",                      inline=True)
        await ctx.reply(embed=embed, mention_author=False)

    # ── !support ──────────────────────────────────────────────────────────────

    @commands.command(name="support", aliases=["help-host"])
    async def support_cmd(self, ctx: commands.Context):
        """Get support information."""
        embed = brand("🎫 KaluxHost Support")
        embed.add_field(name="📩 Open a Ticket",  value="Use `!ticket <issue>` or `/ticket`",  inline=False)
        embed.add_field(name="💬 Discord",         value="Message a staff member",              inline=True)
        embed.add_field(name="📧 Email",            value="support@kaluxhost.com",              inline=True)
        embed.add_field(name="⏱️ Response Time",   value="Usually within 1 hour",              inline=True)
        await ctx.reply(embed=embed, mention_author=False)

    # ── !ticket ───────────────────────────────────────────────────────────────

    @commands.command(name="ticket")
    @commands.guild_only()
    async def ticket_cmd(self, ctx: commands.Context, *, issue: str):
        """Open a support ticket."""
        ticket_embed = discord.Embed(
            title="🎫 New Support Ticket",
            description=f"**Issue:** {issue}",
            color=COLOR_BRAND,
            timestamp=datetime.datetime.now(datetime.timezone.utc),
        )
        ticket_embed.add_field(name="User",    value=f"{ctx.author.mention} (`{ctx.author.id}`)", inline=True)
        ticket_embed.add_field(name="Channel", value=ctx.channel.mention,                         inline=True)
        ticket_embed.add_field(name="Status",  value="🟡 Pending",                                inline=True)
        ticket_embed.set_footer(text="KaluxHost Support | Staff will assist you shortly.")

        await ctx.reply(embed=success("Your support ticket has been submitted! Staff will assist you shortly."), mention_author=False)

        log_ch = discord.utils.find(
            lambda c: c.name in ("ticket-log", "tickets", "mod-log", "support-log"),
            ctx.guild.text_channels,
        )
        if log_ch:
            await log_ch.send(embed=ticket_embed)

    # ── !node ─────────────────────────────────────────────────────────────────

    @commands.command(name="node", aliases=["nodes", "network"])
    async def node_cmd(self, ctx: commands.Context):
        """Show KaluxHost node/network information."""
        embed = brand("🌐 KaluxHost Network")
        nodes = [
            ("🇺🇸 US-East",    "New York",       "Online"),
            ("🇺🇸 US-West",    "Los Angeles",    "Online"),
            ("🇩🇪 EU-Central", "Frankfurt",      "Online"),
            ("🇸🇬 AP-Southeast","Singapore",     "Online"),
        ]
        for flag, loc, status in nodes:
            embed.add_field(name=f"{flag} {loc}", value=f"Status: 🟢 {status}", inline=True)
        await ctx.reply(embed=embed, mention_author=False)

    # ── Slash Commands ────────────────────────────────────────────────────────

    @app_commands.command(name="plans", description="View KaluxHost hosting plans")
    async def slash_plans(self, interaction: discord.Interaction):
        await interaction.response.send_message(embed=_plans_embed())

    @app_commands.command(name="status", description="Check KaluxHost service status")
    async def slash_status(self, interaction: discord.Interaction):
        await interaction.response.send_message(embed=_status_embed())

    @app_commands.command(name="uptime", description="Check bot uptime and latency")
    async def slash_uptime(self, interaction: discord.Interaction):
        embed = brand("📡 Uptime Status")
        embed.add_field(name="🤖 Bot Online Since", value=f"<t:{int(self.bot.start_time.timestamp())}:R>", inline=True)
        embed.add_field(name="📶 API Latency",       value=f"`{round(self.bot.latency*1000)}ms`",           inline=True)
        embed.add_field(name="🟢 Services",           value="All systems operational",                      inline=True)
        await interaction.response.send_message(embed=embed)

    @app_commands.command(name="ticket", description="Open a support ticket")
    @app_commands.guild_only()
    async def slash_ticket(self, interaction: discord.Interaction, issue: str):
        ticket_embed = discord.Embed(
            title="🎫 New Support Ticket",
            description=f"**Issue:** {issue}",
            color=COLOR_BRAND,
            timestamp=datetime.datetime.now(datetime.timezone.utc),
        )
        ticket_embed.add_field(name="User",   value=f"{interaction.user.mention} (`{interaction.user.id}`)", inline=True)
        ticket_embed.add_field(name="Status", value="🟡 Pending",                                            inline=True)
        ticket_embed.set_footer(text="KaluxHost Support | Staff will assist you shortly.")

        await interaction.response.send_message(embed=success("Ticket submitted! Staff will assist you shortly."), ephemeral=True)

        log_ch = discord.utils.find(
            lambda c: c.name in ("ticket-log", "tickets", "mod-log", "support-log"),
            interaction.guild.text_channels,
        )
        if log_ch:
            await log_ch.send(embed=ticket_embed)


async def setup(bot: commands.Bot):
    # Track start time for uptime command
    import datetime
    bot.start_time = getattr(bot, "start_time", datetime.datetime.now(datetime.timezone.utc))
    await bot.add_cog(Hosting(bot))

File: artifacts/discord-bot/src/cogs/hosting.js
/**
 * HOSTING COG — KaluxHost Themed
 * Commands: plans, uptime, status, support, ticket
 * Supports both prefix and slash commands
 */

import { SlashCommandBuilder, EmbedBuilder } from 'discord.js';
import { brandEmbed, successEmbed } from '../utils/embed.js';

const BRAND_COLOR = 0x5865F2;

// ─── Plans Data ───────────────────────────────────────────────────────────────

const PLANS = [
  { name: '⚡ Starter', ram: '2GB', cpu: '1 vCore', storage: '20GB SSD', price: '$2.99/mo', color: 0x57F287 },
  { name: '🚀 Pro', ram: '4GB', cpu: '2 vCores', storage: '50GB SSD', price: '$5.99/mo', color: 0x5865F2 },
  { name: '💎 Elite', ram: '8GB', cpu: '4 vCores', storage: '100GB SSD', price: '$9.99/mo', color: 0xFEE75C },
  { name: '👑 Dedicated', ram: '16GB+', cpu: '8+ vCores', storage: '250GB SSD', price: 'Custom', color: 0xED4245 },
];

// ─── Prefix Commands ──────────────────────────────────────────────────────────

const plansCommand = {
  name: 'plans',
  aliases: ['pricing', 'packages'],
  description: 'View KaluxHost hosting plans',
  cog: 'Hosting',

  async execute(message) {
    const embed = new EmbedBuilder()
      .setColor(BRAND_COLOR)
      .setTitle('🖥️ KaluxHost — Hosting Plans')
      .setDescription('Premium hosting with 99.9% uptime guarantee. All plans include DDoS protection & 24/7 support.')
      .setFooter({ text: 'KaluxHost | Visit our website to order!' })
      .setTimestamp();

    for (const plan of PLANS) {
      embed.addFields({
        name: plan.name,
        value: `RAM: \`${plan.ram}\`\nCPU: \`${plan.cpu}\`\nStorage: \`${plan.storage}\`\nPrice: **${plan.price}**`,
        inline: true,
      });
    }

    await message.reply({ embeds: [embed] });
  },
};

const uptimeCommand = {
  name: 'uptime',
  description: 'Check bot / service uptime',
  cog: 'Hosting',

  async execute(message, _args, client) {
    const ms = client.uptime;
    const embed = brandEmbed('📡 Uptime Status')
      .addFields(
        { name: '🤖 Bot Uptime', value: formatUptime(ms), inline: true },
        { name: '🟢 Service Status', value: 'All systems operational', inline: true },
        { name: '📶 API Latency', value: `${client.ws.ping}ms`, inline: true },
      );
    await message.reply({ embeds: [embed] });
  },
};

const statusCommand = {
  name: 'status',
  description: 'Check KaluxHost service status',
  cog: 'Hosting',

  async execute(message) {
    const services = [
      { name: 'Web Panel', status: '🟢 Online' },
      { name: 'Game Servers', status: '🟢 Online' },
      { name: 'VPS Nodes', status: '🟢 Online' },
      { name: 'Discord Bot', status: '🟢 Online' },
      { name: 'Billing System', status: '🟢 Online' },
    ];

    const embed = brandEmbed('📊 KaluxHost — Service Status')
      .setDescription(services.map((s) => `${s.status} **${s.name}**`).join('\n'));

    await message.reply({ embeds: [embed] });
  },
};

const supportCommand = {
  name: 'support',
  aliases: ['help-hosting'],
  description: 'Get support information',
  cog: 'Hosting',

  async execute(message) {
    const embed = brandEmbed('🎫 KaluxHost Support')
      .addFields(
        { name: '📩 Open a Ticket', value: 'Use `!ticket <issue>` or `/ticket` to open a support ticket', inline: false },
        { name: '💬 Live Chat', value: 'Check our website for live chat support', inline: true },
        { name: '📧 Email', value: 'support@kaluxhost.com', inline: true },
        { name: '⏱️ Response Time', value: 'Usually within 1 hour', inline: true },
      );
    await message.reply({ embeds: [embed] });
  },
};

const ticketCommand = {
  name: 'ticket',
  description: 'Open a support ticket',
  cog: 'Hosting',

  async execute(message, args) {
    if (!args.length)
      return message.reply({ embeds: [brandEmbed('Usage', '`!ticket <describe your issue>`')] });

    const issue = args.join(' ');
    const embed = new EmbedBuilder()
      .setColor(0x5865F2)
      .setTitle('🎫 Support Ticket Created')
      .setDescription(`**Issue:** ${issue}`)
      .addFields(
        { name: 'Submitted by', value: `${message.author.tag}`, inline: true },
        { name: 'User ID', value: message.author.id, inline: true },
        { name: 'Status', value: '🟡 Pending', inline: true },
      )
      .setFooter({ text: 'KaluxHost Support | A staff member will assist you shortly.' })
      .setTimestamp();

    await message.reply({ embeds: [successEmbed('Your support ticket has been submitted! Staff will assist you shortly.') ] });

    // Post to a ticket log channel if one exists named "ticket-log" or "mod-log"
    const logChannel = message.guild.channels.cache.find(
      (c) => ['ticket-log', 'mod-log', 'tickets'].includes(c.name),
    );
    if (logChannel) await logChannel.send({ embeds: [embed] });
  },
};

// ─── Slash Commands ───────────────────────────────────────────────────────────

const slashPlans = {
  data: new SlashCommandBuilder()
    .setName('plans')
    .setDescription('View KaluxHost hosting plans'),

  async execute(interaction) {
    const embed = new EmbedBuilder()
      .setColor(BRAND_COLOR)
      .setTitle('🖥️ KaluxHost — Hosting Plans')
      .setDescription('Premium hosting with 99.9% uptime guarantee. All plans include DDoS protection & 24/7 support.')
      .setFooter({ text: 'KaluxHost | Visit our website to order!' })
      .setTimestamp();

    for (const plan of PLANS) {
      embed.addFields({
        name: plan.name,
        value: `RAM: \`${plan.ram}\`\nCPU: \`${plan.cpu}\`\nStorage: \`${plan.storage}\`\nPrice: **${plan.price}**`,
        inline: true,
      });
    }

    await interaction.reply({ embeds: [embed] });
  },
};

const slashStatus = {
  data: new SlashCommandBuilder()
    .setName('status')
    .setDescription('Check KaluxHost service status'),

  async execute(interaction) {
    const services = [
      { name: 'Web Panel', status: '🟢 Online' },
      { name: 'Game Servers', status: '🟢 Online' },
      { name: 'VPS Nodes', status: '🟢 Online' },
      { name: 'Discord Bot', status: '🟢 Online' },
      { name: 'Billing System', status: '🟢 Online' },
    ];

    const embed = brandEmbed('📊 KaluxHost — Service Status')
      .setDescription(services.map((s) => `${s.status} **${s.name}**`).join('\n'));

    await interaction.reply({ embeds: [embed] });
  },
};

const slashTicket = {
  data: new SlashCommandBuilder()
    .setName('ticket')
    .setDescription('Open a support ticket')
    .addStringOption((o) =>
      o.setName('issue').setDescription('Describe your issue').setRequired(true),
    ),

  async execute(interaction) {
    const issue = interaction.options.getString('issue');
    const embed = new EmbedBuilder()
      .setColor(0x5865F2)
      .setTitle('🎫 Support Ticket Created')
      .setDescription(`**Issue:** ${issue}`)
      .addFields(
        { name: 'Submitted by', value: `${interaction.user.tag}`, inline: true },
        { name: 'User ID', value: interaction.user.id, inline: true },
        { name: 'Status', value: '🟡 Pending', inline: true },
      )
      .setFooter({ text: 'KaluxHost Support | A staff member will assist you shortly.' })
      .setTimestamp();

    await interaction.reply({ embeds: [successEmbed('Your ticket has been submitted! Staff will assist you shortly.')], ephemeral: true });

    const logChannel = interaction.guild.channels.cache.find(
      (c) => ['ticket-log', 'mod-log', 'tickets'].includes(c.name),
    );
    if (logChannel) await logChannel.send({ embeds: [embed] });
  },
};

const slashUptime = {
  data: new SlashCommandBuilder()
    .setName('uptime')
    .setDescription('Check bot uptime'),

  async execute(interaction, client) {
    const embed = brandEmbed('📡 Uptime Status')
      .addFields(
        { name: '🤖 Bot Uptime', value: formatUptime(client.uptime), inline: true },
        { name: '🟢 Service Status', value: 'All systems operational', inline: true },
        { name: '📶 API Latency', value: `${client.ws.ping}ms`, inline: true },
      );
    await interaction.reply({ embeds: [embed] });
  },
};

// ─── Helpers ──────────────────────────────────────────────────────────────────

function formatUptime(ms) {
  const s = Math.floor(ms / 1000);
  const m = Math.floor(s / 60);
  const h = Math.floor(m / 60);
  const d = Math.floor(h / 24);
  return `${d}d ${h % 24}h ${m % 60}m ${s % 60}s`;
}

// ─── Setup ────────────────────────────────────────────────────────────────────

export default {
  setup(client) {
    const prefixCmds = [plansCommand, uptimeCommand, statusCommand, supportCommand, ticketCommand];
    const slashCmds = [slashPlans, slashStatus, slashTicket, slashUptime];

    for (const cmd of prefixCmds) client.commands.set(cmd.name, cmd);
    for (const cmd of slashCmds) client.slashCommands.set(cmd.data.name, cmd);
  },
};
